src/csharp/Microsoft.Spark/Utils/PythonSerDe.cs

```diff
@@ -67,7 +68,23 @@ internal static object[] GetUnpickledObjects(Stream stream, int messageLength)
                     stackCapacity: 102); // Spark sends batches of 100 rows, and +2 is for markers.
                 s_rowConstructor.Reset();
                 Debug.Assert(unpickledItems != null);
-                return (unpickledItems as object[]);
+                // Cast unpickledItems if necessary.
+                foreach (object objArr in (object[])unpickledItems)
+                {
+                    if (objArr.GetType() == typeof(object[]))
+                    {
+                        object obj = ((object[])objArr)[0];
+                        if (obj == null)
+                        {
+                            continue;
+                        }
+                        if (obj.GetType() == typeof(ArrayList))
+                        {
+                            return CastUnpickledItems.UnpickleArray(unpickledItems);
```

Comment:
if you just return here, what will happen to the rest of rows? Would that be ok?

Reply:
I think it should be okay, since `UnpickleArray` takes `unpickledItems` and within the func itself, it will deal with all the rows when it's `ArrayList`.

Reply:
If you get `ArrayList`, all the objects in `unpickledItems` will be a type of `ArrayList`?

Reply:
Yes, from what I have observed. It will look like below:
![image](https://user-images.githubusercontent.com/47404285/91097238-3029c680-e614-11ea-8eeb-46709d525168.png)

Reply:
Maybe we should change this function to `IEnumerable<object> GetUnpickledObjects` and use `yield return`.

Reply:
Do you mean something like below?
```csharp
        internal static IEnumerable<object> GetUnpickledObjects(Stream stream, int messageLength)
        {
            byte[] buffer = ArrayPool<byte>.Shared.Rent(messageLength);

            try
            { 
                ...
                // Check if unpickler returns ArrayList.
                // If so, it needs to be cast to the appropriate array type using CastToArray.
                foreach (object objArr in (object[])unpickledItems)
                {
                    if (objArr.GetType() == typeof(object[]))
                    {
                        object obj = ((object[])objArr)[0];
                        if (obj == null)
                        {
                            continue;
                        }
                        if (obj.GetType() == typeof(ArrayList))
                        {
                            yield return CastUnpickledItems.CastToArray(unpickledItems);
                        }
                        else
                        {
                            yield return unpickledItems;
                        }
                    }
                    else
                    {
                        yield return unpickledItems;
                    }
                }
            }
            finally
            {
                ArrayPool<byte>.Shared.Return(buffer);
            }
        }
```

Reply:
I agree with @imback82 .  You are iterating through all the object[] entries here, and the consumer of `GetUnpickledObjects` will also end up iterating through the object[] again.  It may be better to make this an `IEnumerable`

Reply:
Sure, will make the change tonight.

Reply:
I have refactored this part. Please take a look, thanks!